add a roadmap for the next 6 months (#1784)
libp2p/go-libp2p: ROADMAP.md

# go-libp2p roadmap Q4’22/Q1’23

```
Date: 2022-10-07
Status: In Progress
Notes: This document is still in review will may be heavily modified based on stakeholder feedback. Please add any feedback or questions in:
https://github.com/libp2p/go-libp2p/issues/1806
```

## 🛣️ Milestones
### 2022

#### Early Q4 (October)
-   A.1⚡Handshakes at the Speed of Light - **[Early Muxer Negotiation](https://github.com/libp2p/specs/issues/426)**

> - E.1 🧪 Future-proof testing - **RTT test for Early Muxer Negotiation**

Reviewer: Per below, I think these should be combined.

### Mid Q4 (November)

> -   F.1 ⏱ Future-proof Benchmarking - **[Using nix-builders](https://github.com/testground/testground/pull/1425)**

Reviewer: I think before we spin up benchmarking, lets get interoperability nailed.  The pieces are in place for that already with libp2p/test-plans. I discussed this more below.

#### End of Q4 (December)
-   B.1 🧠 Smart Dialing - **Happy Eyeballs**
-   C 📊 Comprehensive Metrics - **Swarm metrics**

> -   D.1 📺 Universal Browser Connectivity - **[WebRTC Browser -> Server](https://github.com/libp2p/go-libp2p/pull/1655)**

Reviewer: I think the ordering should be:

-   D.1 📺 Universal Browser Connectivity - **[WebRTC Browser -> Server](https://github.com/libp2p/go-libp2p/pull/1655)**
-   B.1 🧠 Smart Dialing - **Happy Eyeballs**
-   C 📊 Comprehensive Metrics - **Swarm metrics**

Reason being:
1. We're not going to do anything to slow down the browser connectivity story.  This closes up the story as well on "all modern browsers can connect with a go-libp2p node". 
2. Smart dialing - assuming we actually hitting a breaking point for what transport to use

Reviewer: Per comment other places I believe, I would bump smart dialing and comprehensive metrics until we're more solid on future proofing interoperability and performance.

### 2023

#### Early Q1 (January)
-   B.2 🧠 Smart Dialing - **QUIC Blackhole detector**
-   A.2 ⚡ Handshakes at the Speed of Light - **[Adding security protocol](https://github.com/libp2p/specs/pull/353)**

#### Mid Q1 (February)
-   B.3 🧠 Smart Dialing - **RTT estimation**
    -   🎉 Estimated Project Completion

#### End of Q1 (March)
-   F.2 ⏱️ Future-proof benchmarking - **Using first class support for remote runners in Testground**
-   A.3 ⚡ Handshakes at the Speed of Light - **0.5 RTT data optimization** (for QUIC)
    -   🎉 Estimated Project Completion

#### Up Next
-   D.2 📺 Universal Browser Connectivity - **[WebTransport](https://github.com/libp2p/go-libp2p/issues/1717)** update to new draft versions
    - Dependency on Chrome needs to support new draft version of WebTransport protocol (still under development by IETF and W3C)
-  E.2 🧪 Future-proof testing - **Comprehensive testing for non-transport features**
-  G 📢 Judicious Address Advertisements

# Roadmap Appendix
## [A.⚡️Handshakes at the Speed of Light](https://github.com/libp2p/go-libp2p/issues/1807)

> **Why**: Historically, libp2p has been very wasteful when it comes to round trips spent during connection establishment. This is slowing down our users, especially their TTFB (time to first byte) metrics.

Reviewer: I agree this is true.  I would love to have more insight on the actual impact.  For example, it would be great to have anecdotes from users on how much this is an adoption blocker.

This is more general comment.  I don't think there's necessarily anything to do right now.  It's more just to push to find ways to quantify problems in terms of impact as much as possible.

I do agree this issue rolls into one of our key pillars of "performance".  TTFB performance isn't the existential threat I have heard complaints about though (vs. libp2p is slower than HTTP).  I could be tuned in here though.  Getting user input would be great.

Reviewer 2: I believe TTFB is an important metric to both Saturn and ipfs cluster folks.

How:

1. Happy Eyeballs: implement some kind of “Happy-Eyeballs” style prioritization among all supported transports
2. QUIC Blackhole detector: detection of blackholes, especially relevant to detect UDP (QUIC) blackholing
3. RTT estimation: estimation of the expected RTT of a connection based on two nodes’ IP addresses, so that Happy Eyeballs Timeouts can be set dynamically

## [C. 📊 Comprehensive Metrics](https://github.com/libp2p/go-libp2p/issues/1356)

**Why**: For far too long, go-libp2p has been a black box. This has hurt us many times, by allowing trivial bugs to go undetected for a long time ([example](https://github.com/ipfs/kubo/pull/8750)). Having metrics will allow us to track the impact of performance improvements we make over time.

**Goal**: Export a wider set of metrics across go-libp2p components and enable node operators to monitor their nodes in production. Optionally provide a sample Grafana dashboard similar to the resource manager dashboard.

**How**: This will look similar to how we already expose resource manager metrics. Metrics can be added incrementally for libp2p’s components. First milestone is having metrics for the swarm.

> ## [D. 📺 Universal Browser Connectivity](https://github.com/libp2p/go-libp2p/issues/1811)

Reviewer: I know the ordering of milestones is defined above. Maybe in a future PR we move the themes around to be in priority order. For example, as a theme, I think browser connectivity should be before all of these. I would likely put future proof testing at the top.

**Why**: A huge part of “the Web” is happening inside the browser. As a universal p2p networking stack, libp2p needs to be able to offer solutions for browser users.

**Goal**: go-libp2p ships with up-to-date WebTransport and (libp2p-) WebRTC implementations, enabled by default. This allows connections between browsers and public nodes, browsers and non-public nodes as well as two browsers.

1. [WebRTC](https://github.com/libp2p/go-libp2p/pull/1655): while browser to public node is getting close to finalized, there’ll be a push to make the other combinations work as well
2. [WebTransport](https://github.com/libp2p/go-libp2p/issues/1717): as the protocol is still under development by IETF and W3C, our implementation needs to follow. To stay up to date, we will have to move as soon as Chrome ships support for a new draft version

## [E. 🧪 Future-proof Testing](https://github.com/libp2p/go-libp2p/issues/1809)

> **Why**: Having lots of transports is great, and shaving off RTTs is awesome. We need to stay backwards-compatible with legacy go-libp2p implementations, and less advanced libp2p stacks.

Reviewer: It's bigger than transports

```suggestion
**Why**: Having lots of transports, encryption mechanism, and stream muxers is great, and shaving off RTTs is awesome. We need to stay backwards-compatible with legacy go-libp2p release, other libp2p implementations, and less advanced libp2p stacks.
```

**Goal**: We have cross-version and cross-implementation Testground coverage that makes sure that we are able to establish a libp2p connection between two nodes, in the expected number of RTTs. This makes sure that the optimizations don’t break compatibility, actually have the performance impact we expect them to have, and serves as a regression test in the future.

> 1. RTT test for Early Muxer Negotiation (A.1)

Reviewer: I'm less convinced this warrants its own milestone.  This is basically part of the done criteria for early muxer integration negotiation.  All new functionality needs to be testing correctly as part of marking it complete.  I think it should be moved up.

I think we need:

Comprehensive interoperability testing across:
libp2p implementations (already happening)
implementation versions (already happening)
transports (I don't think we're exhaustive)
security/encryption (I don't think we're exhaustive)
muxers (I don't think we're exhaustive)

That is the key thing we want I believe so we can move forward with confidence that we're not breaking anything.

> # go-libp2p roadmap Q4’22/Q1’23

Reviewer 3: Great write up! 
A general thought is that as libp2p is getting more widely deployed maybe we also need to think about beef up more security features.  May or may not fit in to the next 6 months time frame.
For example behavior based security. The peer is usually not backed by a root of trust, so the traditional identity and chain of trust based approach does not work in this scenario. We may need to consider new approaches to enforce security.

Author: @julian88110 Do you have anything specific in mind?
Would that belong into the go-libp2p roadmap, or would that first and foremost be an item for the general libp2p roadmap (https://github.com/libp2p/specs/blob/master/ROADMAP.md)?

Reviewer 3: Yes this should belong to the general roadmap for libp2p. I have some thoughts, may take some time to write it up after the muxer thing is done.

## 📊 Comprehensive Metrics

Why: For far too long, go-libp2p has been a black box. This has hurt us many times, by allowing trivial bugs to go undetected for a long time ([example](https://github.com/ipfs/kubo/pull/8750)). Having metrics will allow us to track the impact of performance improvements we make over time.

> Goal: go-libp2p ships with Grafana dashboards that allow node operators to monitor their nodes in production.

Reviewer 4: ```suggestion
Goal:  Export a wider set of metrics across go-libp2p components and enable node operators to monitor their nodes in production. Optionally provide a sample Grafana dashboards similar to the [rust-libp2p dashboard](https://kademlia-exporter.max-inden.de/d/Pfr0Fj6Mk/rust-libp2p?orgId=1&refresh=30s)
```

Author: Resolved by pointing to our rcmgr Grafana dashborad.

> Goal: go-libp2p ships with Grafana dashboards that allow node operators to monitor their nodes in production.

Author: ```suggestion
Goal: Export a wider set of metrics across go-libp2p components and enable node operators to monitor their nodes in production. Optionally provide a sample Grafana dashboard similar to the resource manager dashboard.
```

> How: This will look similar to how we already expose resource manager metrics. Metrics can be added incrementally for libp2p’s components. First milestone is having metrics for the swarm.

Reviewer 4: Is the goal here to actually check in a Grafana dashboard json model to this repo and direct people to use that for monitoring? or only to slowly add metrics to go-libp2p components and enable users of go-libp2p to craft their own Grafana dashboards

Author: The goal is to have a ready-to-use Grafana dashboard _somewhere_. Of course, users are free to ignore our dashboards and build whatever they like from the metrics we expose.

Not sure if checking in the json file here is the most user-friendly way, especially as the number of json files will grow quickly. I don't think we need to figure out the best way to publish those right now, that can be part of the work we're planning to do.

Reviewer 4: Yeah that makes sense. Other projects usually have community members that create their own dashboards for use by others. I'm thinking of something like this: https://github.com/dguess/mina-node-dashboard

> Not sure if checking in the json file here is the most user-friendly way

I agree, "go-libp2p ships with Grafana dashboards" just threw me off. Thanks for the clarification.

Author: Any suggestion how to rephrase?

Reviewer 4: I added my suggestion https://github.com/libp2p/go-libp2p/pull/1784/files#r981794624

Author: Resolving this thread, since I've reworded the goal to be more specific.

2. WebRTC: while browser to public node is getting close to finalized, there’ll be a push to make the other combinations work as well

## ⚡️Handshakes at the Speed of Light

Why: Historically, libp2p has been very wasteful when it comes to round trips spent during connection establishment. This is slowing down our users, especially their TTFB (time to first byte) metrics.

Goal: go-libp2p optimizes its handshake latency up to the point where only increasing the speed of light would lead to further speedups. In particular, this means (in chronological order of a handshake):

1. cutting off the 1 RTT wasted on security protocol negotiation by including the security protocol in the multiaddr: [https://github.com/libp2p/specs/pull/353](https://github.com/libp2p/specs/pull/353)
2. cutting off the 1 RTT wasted on muxer negotiation: [https://github.com/libp2p/specs/issues/426](https://github.com/libp2p/specs/issues/426)
3. using 0.5-RTT data (for TLS) / a Noise Extension to ship the list of Identify protocols, cutting of 1 RTT that many protocols spend waiting on `IdentifyWait`

## 🧠 Smart Dialing

> Why: Having a large list of transports to pick from is great. Having an advanced stack that can dial all of them is even greater. But dialing all of them at the same time wastes our, the network’s and the peer’s resources. 

Reviewer 5: ```suggestion
Why: Having a large list of transports to pick from is great. Having an advanced stack that can dial all of them is even greater. But dialing all of them at the same time wastes the network’s and the peer’s resources. 
```

Author: It also wastes our CPU cycles to start that many handshakes.

Goal: When given a list of multiaddrs of a peer, go-libp2p is smart enough to pick the address that results in the most performant connection (for example, preferring QUIC over TCP), while also picking the address such that maximizes the likelihood of a successful handshake.

How:

1. implement some kind of “Happy-Eyeballs” style prioritization among all supported transports
2. estimation of the expected RTT of a connection based on two nodes’ IP addresses, so that Happy Eyeballs Timeouts can be set dynamically
3. detection of blackholes, especially relevant to detect UDP (QUIC) blackholing

## 🧪 Future-proof Testing

Why: Having lots of transports is great, and shaving off RTTs is awesome. We need to stay backwards-compatible with legacy go-libp2p implementations, and less advanced libp2p stacks.

> Goal: We have cross-version and cross-implementation Testground coverage that makes sure that we are able to establish a libp2p connection between two nodes, in the expected number of RTTs. This makes sure that the optimizations don’t break compatibility, actually have the performance impact we expect them to have, and serves as a regression test in the future.

Reviewer 3: Also thinking about adding the item to convert time-delay based test cases into event sync based test cases if it fits into the time frame, that will reduce a lot of the flakiness.

Author: Definitely worth doing. That would probably be part of the general project maintenance / tech debt point that @p-shahi was asking for in https://github.com/libp2p/go-libp2p/pull/1784#issuecomment-1259955399.

1. cutting off the 1 RTT wasted on security protocol negotiation by including the security protocol in the multiaddr: [https://github.com/libp2p/specs/pull/353](https://github.com/libp2p/specs/pull/353)
2. cutting off the 1 RTT wasted on muxer negotiation: [https://github.com/libp2p/specs/issues/426](https://github.com/libp2p/specs/issues/426)
3. using 0.5-RTT data (for TLS) / a Noise Extension to ship the list of Identify protocols, cutting of 1 RTT that many protocols spend waiting on `IdentifyWait`

## 🧠 Smart Dialing

Why: Having a large list of transports to pick from is great. Having an advanced stack that can dial all of them is even greater. But dialing all of them at the same time wastes our, the network’s and the peer’s resources. 

Goal: When given a list of multiaddrs of a peer, go-libp2p is smart enough to pick the address that results in the most performant connection (for example, preferring QUIC over TCP), while also picking the address such that maximizes the likelihood of a successful handshake.

How:

> 1. implement some kind of “Happy-Eyeballs” style prioritization among all supported transports

Reviewer: ```suggestion
1. implement some kind of “[Happy-Eyeballs](https://en.wikipedia.org/wiki/Happy_Eyeballs)” style prioritization among all supported transports
```

2. estimation of the expected RTT of a connection based on two nodes’ IP addresses, so that Happy Eyeballs Timeouts can be set dynamically
3. detection of blackholes, especially relevant to detect UDP (QUIC) blackholing

## 🧪 Future-proof Testing

Why: Having lots of transports is great, and shaving off RTTs is awesome. We need to stay backwards-compatible with legacy go-libp2p implementations, and less advanced libp2p stacks.

> Goal: We have cross-version and cross-implementation Testground coverage that makes sure that we are able to establish a libp2p connection between two nodes, in the expected number of RTTs. This makes sure that the optimizations don’t break compatibility, actually have the performance impact we expect them to have, and serves as a regression test in the future.

Reviewer: We talk about ensuring compability between versions and implementations, but I assume there is an extra dimension we need to add around transports (and likely encryption and muxer).

Basically want to show 

<go-libp2p, vX, QUIC> can talk with <go-libp2p, vY, QUIC>
<go-libp2p, vX, TCP> can talk with <go-libp2p, vY, TCP>
<go-libp2p, vX, WebTransport> can talk with <go-libp2p, vY, WebTranport>
etc.
